packages/itwinui-react/src/core/Menu/Menu.tsx

```diff
@@ -49,67 +59,45 @@ export const Menu = React.forwardRef((props, ref) => {
   }, []);
 
   React.useEffect(() => {
```

Comment:
You could probably use a callback `ref` instead of `useEffect`, but you would still need `getFocusableNodes`.

However, the main thing that this code is trying to solve is based on a flawed understanding of how the component is supposed to behave. Ideally, there would only be `MenuItem`s inside the `Menu`. But currently we support `MenuExtraContent` with any arbitrary focusable elements, which is not at all how a menu is supposed to work.

How difficult would it be to make it such that it uses regular <kbd>Tab</kbd> navigation for `MenuExtraContent`, and then only uses arrow-key navigation for the actual menu-items? Would likely need a wrapper around the menu-items (could be automatically added maybe?):

```jsx
<Menu>
  <MenuExtraContent>
    <button>tab to me</button>
  </MenuExtraContent>
  <Divider />
  <MenuList>
    <MenuItem>tab again</MenuItem>
    <MenuItem>now use arrow keys</MenuItem>
    <MenuItem>more arrows 🏹</MenuItem>
  </MenuList>
</Menu>
```

Additionally, we should probably deprecate `MenuExtraContent` in favor of `Popover`.

Reply:
> How difficult would it be to make it such that it uses regular Tab navigation for `MenuExtraContent`, and then only uses arrow-key navigation for the actual menu-items?

I believe it would take a bit of thinking to get all aspects right and not introduce any regressions. So, I feel a simple solution that keeps the same behavior might be more suitable here, at least for this PR.

I played around with using `useSyncExternalStore` instead of the `useEffect` ([code](https://github.com/iTwin/iTwinUI/pull/2021/files#diff-eb1d53cfaf5ebd716d4feb107991cc2f5775bf2d4e9cc83b9de60e2206dea595R184-R208)). I think it works for most of the cases. Will be testing it more thoroughly.

Reply:
> I played around with using `useSyncExternalStore` instead of the `useEffect` ([code](https://github.com/iTwin/iTwinUI/pull/2021/files#diff-eb1d53cfaf5ebd716d4feb107991cc2f5775bf2d4e9cc83b9de60e2206dea595R184-R208)). I think it works for most of the cases. Will be testing it more thoroughly.

After properly applying `getItemProps` (https://github.com/iTwin/iTwinUI/pull/2021#discussion_r1612095985), I realized that the extra `useSyncExternalStore` to keep `useListNavigation`'s activeIndex in sync whenever an item was hovered/focused was no longer needed. I, thus, removed it (https://github.com/iTwin/iTwinUI/pull/2021/commits/f75b3df2afa4f35f0a565e9ff2c476259f4a9eca).

However, I still could not remove the `useEffect`. If we have plans to only support `MenuItem`s and probably deprecate `MenuExtraContent`, I guess this `useEffect` is alright since it will be removed in the future.